### ISSUE TITLE
Setting fcitx-use-dbus to t leads to command lag

### ISSUE THREAD

**Issue by et2010** [cute-jumper/fcitx.el#30]
For detail, see here:
https://emacs-china.org/t/debug-emacs-performance-issue-ivy/2448/7
For short:
Some serious lagging issue is caused by that option. I'm using spacemacs on Linux. When I turned that option to t, double tap on SPC will lag for 2 seconds before candidates list shows. When turned off, everything including fcitx.el works fine.

So I think I should raise an alarm here in case anyone else has the same issue.

**Comment by cute-jumper** [comment]
Yep. I can confirm that setting `fcitx-use-dbus` to `t` causes "lagging" in **echo area**. It makes the previous message stay in the echo area for a while. Since `ivy` uses minibuffer, it could be a problem.

I am not sure why using dbus can cause the problem for the time being. I would suggest not setting that variable for now if you're using `ivy`.

**Comment by cute-jumper** [comment]
From what I can tell, this is caused by `echo-keystrokes` and `dbus`. Not sure what exact reason that leads to this problem. If you set `echo-keystrokes` to a larger number, say 1, or totally disable echoing keystrokes by setting it to 0, the problem seems to go. It could be a bug in `dbus-call-method`.

BTW, what is your `echo-keystrokes` value? Does the method I mentioned above work for you?

**Comment by et2010** [comment]
Yes, my `echo-keystrokes` is 0.02, and setting it to 0 or 1 works for me.

But I have some questions. Should I enable `fcitx-use-dbus` option on Linux? What would I miss if I set it to nil? It seems that it works just fine even without that option.

**Comment by cute-jumper** [comment]
> Should I enable fcitx-use-dbus option on Linux in the first place? 
No. This is **just** an optimization. Using `dbus` is faster in Linux, although the difference is not that big. You could choose not to set the variable to `t`, which is totally fine.

**Comment by et2010** [comment]
Thanks for the clarification. I guess I'll keep that option and set `echo-keystrokes` to 1.  Is there any side effects for doing this? I don't know anything about `echo-keystrokes` either except for what its docstring says which means little for me. I just want to make sure I don't break something else this time 😂

**Comment by cute-jumper** [comment]
`echo-keystrokes` controls the delay between you press a key and the key shows in the echo area. So if you set it to 1, you need to wait for one second to see your previous keystroke in the echo area. If that's unacceptable, you don't need to set `fcitx-use-dbus` to `t`. Otherwise it would be fine.

The performance improvement of using `dbus` depends on your machine. You can measure the running time of `fcitx--active-p-proc` and `fcitx--active-p-dbus` (and there are other two pairs of functions with similar names) on your own machine to see the difference. Setting `fcitx-use-dbus` to `t` will use the functions with `-dbus` suffix. Whether to use `dbus` or not is just a personal preference.

I would update the README about `fcitx-use-dbus` to make people aware of the problem.

**Comment by et2010** [comment]
Yes, dbus option is faster. On my computer, running:
```elisp
(benchmark-run 100 (fcitx--active-p-proc))
(benchmark-run 100 (fcitx--active-p-dbus))
```
yeilds
```
(0.343457615 0 0.0)
(0.202333871 0 0.0)
```
I'll keep my setting for now.
